hashing.go

```diff
@@ -0,0 +1,170 @@
+		block := blocks[0]
+		return func(seed maphash.Seed, v T) uint64 {
+			return maphash.Bytes(seed, asBytes(unsafe.Pointer(&v), block.Offset, block.Length))
+		}
+	}
+
+	// data type is a single contiguous region of memory (e.g. struct with integer fields)
+	if len(blocks) == 1 && blocks[0].IsString {
+		block := blocks[0]
+		return func(seed maphash.Seed, v T) uint64 {
+			return maphash.String(seed, asString(unsafe.Pointer(&v), block.Offset))
+		}
+	}
+
+	// complex data type consisting of multiple blocks
+	{
+		return func(seed maphash.Seed, v T) uint64 {
+			var h maphash.Hash
+			h.SetSeed(seed)
+			ptr := unsafe.Pointer(&v)
+
+			for _, b := range blocks {
+				if b.IsString {
+					h.WriteString(asString(ptr, b.Offset))
+				} else {
+					h.Write(asBytes(ptr, b.Offset, b.Length))
+				}
+			}
+			return h.Sum64()
+		}
+	}
+}
+
+// asString interprets memory at p + offset as a string header and returns the string.
+func asString(p unsafe.Pointer, offset uintptr) string {
+	sp := unsafe.Add(p, offset)
+	return *(*string)(sp)
+}
+
+// asBytes returns block of memory at p + offset as a byte slice.
+func asBytes(p unsafe.Pointer, offset uintptr, length uintptr) []byte {
+	sp := unsafe.Add(p, offset)
+
+	// Fast approach: 4.5 ns/op for int hashing.
+	// Dirty, because runtime thinks that returned slice hash 1Tb of capacity
+	return (*[1 << 40]byte)(sp)[:length]
+
+	// go 1.20 way: 4.5 ns/op for int hashing
+	// Need to change go.mod
+	//return unsafe.Slice((*byte)(sp), int(length))
+
+	// Clean pre go 1.20 way: 12 ns/op for int hashing
+	// Very slow compared to others
+	//var res []byte
+	//sh := (*reflect.SliceHeader)(unsafe.Pointer(&res))
+	//sh.Data = uintptr(sp)
+	//sh.Len = int(length)
+	//sh.Cap = int(length)
+	//return res
+}
+
+type hMemBlock struct {
+	Offset   uintptr
+	Length   uintptr
+	IsString bool
+}
+
+// hGetMemLayout returns memory layout of the given type.
+// It works for comparable types with any level of nesting as long as there's no interfaces anywhere inside.
+// Each type can be described as a list of memory blocks, where each block is either a string or a plain block.
+// For example:
+// - Point3d(x,y,z) is a single plain block of size 24
+// - User(id, name, age, height) is 3 blocks plain(id), string(name), plain(age, height)
+// Type can consist of multiple blocks even when there's no strings, but when padding is involved.
+func hGetMemLayout(t reflect.Type) []hMemBlock {
+	if !t.Comparable() {
+		// this filters out slices, maps, functions, etc
+		panic("type is not comparable")
+	}
+
+	return getGetMemLayoutAcc(t, 0, nil)
+}
+
+func getGetMemLayoutAcc(t reflect.Type, offset uintptr, acc []hMemBlock) []hMemBlock {
+	switch t.Kind() {
+	case reflect.Struct:
+		n := t.NumField()
+		for i := 0; i < n; i++ {
+			f := t.Field(i)
+			if f.Name == "_" {
+				continue
+			}
+			acc = getGetMemLayoutAcc(f.Type, offset+f.Offset, acc)
+		}
+
+	case reflect.Array:
+		// Optional optimization for [n]byte.
```

Comment:
Do we really need to optimize for byte array? It shouldn't be `Comparable` unless I'm confusing something.

Reply:
Array are comparable, they're fixed-size. Slice aren't. As far as I remember everything is comparable except slices, maps and functions.

Optimization for [n]byte is optional. But taking into account your previous comment, I don't know anymore how to it right. Say we have array of integers, should we hash each integer separately, or hash all of them as contiguous memory block?

Reply:
> Array are comparable, they're fixed-size.

Thanks, I keep confusing this all the time. Strings are comparable while other var-size types aren't. :(